google/cloud/aiplatform/training_jobs.py

```diff
@@ -66,6 +66,288 @@
+                point to a location where the user only has a
+                read access.
+            training_task_inputs (~.struct.Value):
+                Required. The training task's parameter(s), as specified in
+                the
+                ``training_task_definition``'s
+                ``inputs``.
+            dataset (aiplatform.Dataset):
+                Optional. The dataset within the same Project from which data will be used to train the Model. The
+                Dataset must use schema compatible with Model being trained,
+                and what is compatible should be described in the used
+                TrainingPipeline's [training_task_definition]
+                [google.cloud.aiplatform.v1beta1.TrainingPipeline.training_task_definition].
+                For tabular Datasets, all their data is exported to
+                training, to pick and choose from.
+            training_fraction_split (float):
+                The fraction of the input data that is to be
+                used to train the Model. This is ignored if Dataset is not provided.
+            validation_fraction_split (float):
+                The fraction of the input data that is to be
+                used to validate the Model. This is ignored if Dataset is not provided.
+            test_fraction_split (float):
+                The fraction of the input data that is to be
+                used to evaluate the Model. This is ignored if Dataset is not provided.
+            model (~.model.Model):
+                Optional. Describes the Model that may be uploaded (via
+                [ModelService.UploadMode][]) by this TrainingPipeline. The
+                TrainingPipeline's
+                ``training_task_definition``
+                should make clear whether this Model description should be
+                populated, and if there are any special requirements
+                regarding how it should be filled. If nothing is mentioned
+                in the
+                ``training_task_definition``,
+                then it should be assumed that this field should not be
+                filled and the training task either uploads the Model
+                without a need of this information, or that training task
+                does not support uploading a Model as part of the pipeline.
+                When the Pipeline's state becomes
+                ``PIPELINE_STATE_SUCCEEDED`` and the trained Model had been
+                uploaded into AI Platform, then the model_to_upload's
+                resource ``name``
+                is populated. The Model is always uploaded into the Project
+                and Location in which this pipeline is.
+            base_output_dir (str):
+                Optional. GCS output directory of job. If not provided a
+                timestamped directory in the staging directory will be used.
+    """
+
+        if self._has_run:
```

Comment:
The issue with calling this here is that the child class, Custom Training, would have already executed logic with side effects that should have been avoided. In this particular instance, the python package would have been created and uploaded to GCS and then we would raise here when we should have raised earlier. This should be moved to the top of the child classes run call.

Reply:
Ok, got it

Reply:
Fixed